sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -0,0 +1,274 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.CatalystConf
+import org.apache.spark.sql.catalyst.expressions.{And, AttributeSet, Expression, PredicateHelper}
+import org.apache.spark.sql.catalyst.plans.{Inner, InnerLike}
+import org.apache.spark.sql.catalyst.plans.logical.{Join, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+
+/**
+ * Cost-based join reorder.
+ * We may have several join reorder algorithms in the future. This class is the entry of these
+ * algorithms, and chooses which one to use.
+ */
+case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!conf.cboEnabled || !conf.joinReorderEnabled) {
+      plan
+    } else {
+      plan transform {
+        case p @ Project(projectList, j @ Join(_, _, _: InnerLike, _)) if !j.ordered =>
+          reorder(j, p.outputSet)
+        case j @ Join(_, _, _: InnerLike, _) if !j.ordered =>
+          reorder(j, j.outputSet)
+      }
+    }
+  }
+
+  def reorder(plan: LogicalPlan, output: AttributeSet): LogicalPlan = {
+    val (items, conditions) = extractInnerJoins(plan)
+    val result =
+      if (items.size > 2 && items.size <= conf.joinReorderDPThreshold && conditions.nonEmpty) {
+        JoinReorderDP(conf, items, conditions, output).search().getOrElse(plan)
+      } else {
+        plan
+      }
+    // Set all inside joins ordered.
+    setOrdered(result)
+    result
+  }
+
+  /**
+   * Extract inner joinable items and join conditions.
+   * This method works for bushy trees and left/right deep trees.
+   */
+  def extractInnerJoins(plan: LogicalPlan): (Seq[LogicalPlan], Set[Expression]) = plan match {
+    case j @ Join(left, right, _: InnerLike, cond) =>
+      val (leftPlans, leftConditions) = extractInnerJoins(left)
+      val (rightPlans, rightConditions) = extractInnerJoins(right)
+      (leftPlans ++ rightPlans, cond.map(splitConjunctivePredicates).getOrElse(Nil).toSet ++
+        leftConditions ++ rightConditions)
+    case Project(_, j @ Join(left, right, _: InnerLike, cond)) =>
+      val (leftPlans, leftConditions) = extractInnerJoins(left)
+      val (rightPlans, rightConditions) = extractInnerJoins(right)
+      (leftPlans ++ rightPlans, cond.map(splitConjunctivePredicates).getOrElse(Nil).toSet ++
+        leftConditions ++ rightConditions)
+    case _ =>
+      (Seq(plan), Set())
+  }
+
+  def setOrdered(plan: LogicalPlan): Unit = plan match {
+    case j @ Join(left, right, _: InnerLike, cond) =>
+      j.ordered = true
+      setOrdered(left)
+      setOrdered(right)
+    case Project(_, j @ Join(left, right, _: InnerLike, cond)) =>
+      j.ordered = true
+      setOrdered(left)
+      setOrdered(right)
+    case _ =>
+  }
+}
+
+/**
+ * Reorder the joins using a dynamic programming algorithm:
```

Comment:
Thanks for the material. 
Usually for small number of joins (say less than 12), the current DP algorithm can work well. It has support from both academy (Selinger's paper) and industry (postgresql's standard join reorder). 
For more complex joins and large number of joins, there will need lots of research and comparison for many kinds of algorithms, e.g. genetic algorithm is used in postgresql.